bluetooth: controller: Implemented dynamic re-scheduling in ticker

mtpr-ot opened the pull request:
As a part of a ticker extension interface, it is now possible to specify
a slot_window when starting a ticker. When setting the
ticks_slot_window to a non-zero value, it is requested that the node
timeout is re-located to a position within the window, where the node
does not collide with other nodes - aligning to the end of the window.
The solution takes into consideration if a node has already been
updated with drift correction (e.g. ADV randomization), subtracting this
from the window.

Fixes #20264

Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -879,7 +886,10 @@ u8_t ll_adv_enable(u8_t enable)
-				   (adv->evt.ticks_slot + ticks_slot_overhead),
+				   ticks_slot,
+#if defined(CONFIG_BT_TICKER_EXT)
+				   HAL_TICKER_US_TO_TICKS(10000) + ticks_slot,
```
Should we have a define for 10000 ?

cvinayak commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -82,18 +93,36 @@ struct ticker_node {
-	u32_t ticks_periodic;		/* Ticker period ticks */
-	u32_t remainder_periodic;	/* Sub-microsecond tick remainder */
-	u16_t lazy;			/* Periodic latency in number of
+	u32_t  ticks_at_start;		/* Anchor ticks (absolute) */
```
Why the extra space added?

mtpr-ot replied:
> Why the extra space added?

Glitch ;-)

cvinayak commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -62,6 +62,17 @@ struct ticker_node {
 					  */
 #if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE)
+#if  defined(CONFIG_BT_TICKER_EXT)
+	u32_t ticks_slot_window;	 /* Window in which the slot
```
Can we get these 9 bytes (12 bytes unpacked) of memory supplied by user of the extension feature? Just have a pointer here to reference the memory defined by user/caller of ticker_start_ext interface.

This way only the advertiser will use 16 bytes extra, where as other states/roles use 4 bytes extra.

mtpr-ot replied:
Good idea

cvinayak commented:
samples/bluetooth/scan_adv, modified to use scan interval of 100ms and scan_window of 50 ms.

Before this PR, advertiser and scanner that overlapped skipped one of the state's event.

![Before_commented](https://user-images.githubusercontent.com/6350656/72605561-4dc7dc80-391d-11ea-89e0-ae73e81f689c.PNG)

In comparison to ticker compatibility mode wherein collision mitigation was based on first-scheduled instance getting higher precedence, in the current ticker implementation current active/expired ticker instance gets precedence (which is not fair), and colliding instance needs to apply priority in order to grab the time space it is colliding with.

After this PR, advertiser pre-empts scanner on collision (without exercising priority based) and does not skip its time space. But on being done with its event, scanner is resumed with fresh scan window, this is not correct. 

![After_commented](https://user-images.githubusercontent.com/6350656/72606359-c24f4b00-391e-11ea-8d39-d3e90e311811.PNG)

Is the new behavior as expected with this PR?

Thinking aloud, this PR should have slid the advertiser in +/- 10 ms window trying to be not skipped. But if it cant, it will have to skip without disturbing the scan window.

mtpr-ot commented:
> In comparison to ticker compatibility mode wherein collision mitigation was based on first-scheduled instance getting higher precedence, in the current ticker implementation current active/expired ticker instance gets precedence (which is not fair), and colliding instance needs to apply priority in order to grab the time space it is colliding with.

According to the current ticker design (not compatibility mode), the expired ticker does not automatically get precedence. The "age" (highest period ticks) weighs equal to latency (lazy), and priority is only used if applied by host. This seems fair. If that's not what you see, I should investigate.

mtpr-ot commented:
> After this PR, advertiser pre-empts scanner on collision (without exercising priority based) and does not skip its time space. But on being done with its event, scanner is resumed with fresh scan window, this is not correct.

No, this is not what I expect. If advertiser cannot find a slot within the 10 ms. window, it "dumps" the node at the end of its window, expecting a collision and being skipped. This is necessary for honoring "must expire". 
It is also supposed to have the advantage that if the node it collided with is updated, there is a chance it wil get air time anyway. If the ADV preempts the scanner, there's seems to be a bug. It should not be able to prempt anything.

cvinayak commented:
> If that's not what you see, I should investigate.

This is not related to this PR, but something that I have noticed when using nRF flash driver that uses ticker at 7.5ms and a master connection establishment at 7.5 ms, that they dont round-robin and connection fails.

> If the ADV preempts the scanner, there seems to be a bug. It should not be able to preempt anything.

Yes, this seems to be a bug in this PR. One, ticker should not expire for the advertiser that overlaps the scanner; two, seem a secondary bug in LLL scan that it resumed with restored window (this is not observed otherwise when ticker does not expire an instance that overlaps scan window).

Hope you can address the issue "One".

mtpr-ot commented:
Found the bug with preemption - the "previous" data was reset incorrectly, thereby forgetting that a node already reserved the slot. @cvinayak, please retest if you can find the time. Thanks.

mtpr-ot commented:
> This is not related to this PR, but something that I have noticed when using nRF flash driver that uses ticker at 7.5ms and a master connection establishment at 7.5 ms, that they dont round-robin and connection fails.

I would like to investigate this. I have run several similar scenarios without problems, but maybe the nRF LLL makes the behavior different from what we see. Any idea how to reproduce in simulation?

cvinayak commented:
> Found the bug with preemption - the "previous" data was reset prematurely, thereby forgetting that a node already reserved the slot.

I have tested this and the PR now does not introduce any regression in scan_adv sample.

mtpr-ot commented:
> I have tested this and the PR now does not introduce any regression in scan_adv sample.

Thanks Vinayak.

mtpr-ot commented:
Just did a DMEM profiling with and without BT_TICKER_EXT feature enabled.
My configuration includes CONFIG_BT_BROADCASTER, CONFIG_BT_OBSERVER and CONFIG_BT_MAX_CONN=5.

The added DMEM usage for BT_TICKER_EXT reported by our target compiler is **92 bytes**.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -791,17 +811,44 @@ void ticker_worker(void *param)
+				ticker->ext_data->reschedule_state =
+					TICKER_RESCHEDULE_STATE_NONE;
+			}
+#endif /* CONFIG_BT_TICKER_EXT */
```
Here a collision has been detected. It is checked if node has slot_window set, and if so set the reschedule state pending for ticker_job.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -791,17 +811,44 @@ void ticker_worker(void *param)
-				/* Not a must-expire case or this is programmed
-				 * latency. Skip this ticker node
+			    (ticker->lazy_periodic >= ticker->lazy_current) ||
+			    TICKER_RESCHEDULE_PENDING(ticker)) {
```
We always skip timeout callback if rescheduling has become pending. Must-expire is handled at rescheduled time- also if there's no free slot found.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1056,6 +1103,14 @@ static inline void ticker_job_node_update(struct ticker_node *ticker,
+			user_op->params.update.ticks_drift_plus -
+			user_op->params.update.ticks_drift_minus;
+	}
+#endif /* CONFIG_BT_TICKER_EXT */
```
Store the applied drift - we must reduce the rescheduling window by this value.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1303,11 +1359,17 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
+						id_expired;
+					instance->ticks_slot_previous =
+						ticker->ticks_slot;
+				}
```
This change to make sure that when rescheduling a node, the id_slot_previous and slot_previous information is not cleared, as we need it for the next ticker_worker.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1317,11 +1379,21 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
+				 * collision resolved after all nodes are
+				 * restarted
+				 */
+				ticker->ticks_to_expire = ticks_elapsed;
```
Just puts the reschedule-pending node in front for rescheduling in ticker_job.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -791,17 +811,44 @@ void ticker_worker(void *param)
 		if (ticker->ticks_slot != 0U &&
 		   (slot_reserved || ticker_resolve_collision(node, ticker))) {
+#if defined(CONFIG_BT_TICKER_EXT)
+			if (ticker->ext_data &&
```
Action: Dereference ticker->ext_data

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1303,11 +1359,17 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
+			/* If a reschedule is set pending, we will need to keep
+			 * the slot_previous information
+			 */
+			if (!TICKER_RESCHEDULE_PENDING(ticker)) {
```
Action: comine into one statement (AND)

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1432,7 +1507,214 @@ static inline u32_t ticker_job_insert(struct ticker_instance *instance,
+		ext_data->reschedule_state =
+			TICKER_RESCHEDULE_STATE_DONE;
+
+		rescheduling = 1U;
```
Action: Remove

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1432,7 +1507,214 @@ static inline u32_t ticker_job_insert(struct ticker_instance *instance,
+		 * taken into consideration. Reduce the window with
+		 * the amount of drift already applied.
+		 */
+		if (ext_data->ticks_drift < ext_data->ticks_slot_window) {
```
Action: Add TODO regarding appending the random delay to the window...

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1432,7 +1507,214 @@ static inline u32_t ticker_job_insert(struct ticker_instance *instance,
+		if (ticker_id_prev != TICKER_NULL) {
+			/* Node did not become the first - insert it */
+			instance->ticker_id_head   = nodes[ticker_id_head].next;
+			nodes[ticker_id_head].next = nodes[ticker_id_prev].next;
```
Action: Add blank line to separate head update with re-linking. Second part comment: "Re-insert/link ticker nodes"

mtpr-ot committed:
bluetooth: controller: Implemented dynamic re-scheduling in ticker

As a part of a ticker extension interface, it is now possible to specify
a slot_window when starting a ticker. When setting the
ticks_slot_window to a non-zero value, it is requested that the node
timeout is re-located to a position within the window, where the node
does not collide with other nodes - aligning to the end of the window.
The solution takes into consideration if a node has already been
updated with drift correction (e.g. ADV randomization), subtracting this
from the window.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot committed:
bluetooth: controller: Added use of slot_window for ADV

Request ticker re-scheduling of ADV (within the 10ms randomized window).

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot commented:
@cvinayak, PR now updated according to actions, please evaluate. Thanks.

cvinayak commented:
scan_adv sample observations (mesh usecase):

Ticker compatibility,
![image](https://user-images.githubusercontent.com/6350656/73189773-3fe83780-4125-11ea-9507-5065d01435e6.png)

Continuous scanning,
![image](https://user-images.githubusercontent.com/6350656/73189925-76be4d80-4125-11ea-98a6-5e8c0b03b6e8.png)

Continuous scanning with PR,
![image](https://user-images.githubusercontent.com/6350656/73189976-90f82b80-4125-11ea-9eb3-d1dc5856dffe.png)

100ms interval and 50 ms window,
![image](https://user-images.githubusercontent.com/6350656/73190021-a1a8a180-4125-11ea-9a99-3b28e696e1dc.png)

100ms interval and 50 ms window with PR,
![image](https://user-images.githubusercontent.com/6350656/73190069-b38a4480-4125-11ea-848f-cf526692f451.png)

Changes in this PR look good.
